lightningd/lightningd.c

```diff
@@ -251,6 +253,38 @@ static void daemonize_but_keep_dir(void)
 	tal_free(cwd);
 }
 
+static void pidfile_create(const struct lightningd *ld)
+{
+        int pid_fd;
+        char *pid;
+        const tal_t *tmpctx = tal_tmpctx(NULL);
+
+        /* Create PID file */
+        pid_fd = open(ld->pidfile, O_WRONLY|O_CREAT, 0640);
+        if (pid_fd < 0)
+                errx(1, "Failed to open PID file: %s", strerror(errno));
+
+        /* Lock PID file */
+        if (lockf(pid_fd, F_TLOCK, 0) < 0)
+                /* Can't lock file */
+                errx(1, "Cannot lock PID file: %s", strerror(errno));
+
+        /* Get current PID */
+        pid = tal_fmt(tmpctx, "%d\n", getpid());
+
+        /* Write PID to lockfile */
+        if (!write_all(pid_fd, pid, strlen(pid)))
+                /* Can't write file */
+                errx(1, "Cannot write PID file: %s", strerror(errno));
+
+        /* Close PID file */
+        if (close(pid_fd) < 0)
+                /* Can't close file */
```

Comment:
This comment too :)

Reply:
And you can't close the file if you were relying on the `flock` to prevent multiple lightningds from hitting it?

Reply:
F_LOCK Set an exclusive lock on the specified section of the file.
              If (part of) this section is already locked, the call blocks
              until the previous lock is released.  If this section overlaps
              an earlier locked section, both are merged.  File locks are
              released as soon as the process holding the locks closes some
              file descriptor for the file.  A child process does not
              inherit these locks.

       F_TLOCK
              Same as F_LOCK but the call never blocks and returns an error
              instead if the file is already locked.